stdlib/pqueue.ml

```diff
@@ -0,0 +1,287 @@
+    let parent_node i = (i - 1) / 2
+
+    (* We say that a heap respects the "heap ordering" if the value of
+       each node is no greater than the value of its children. The
+       algorithm manipulates arrays that respect the heap algorithm,
+       except for one node whose value may be too small or too large.
+
+       The auxiliary functions [sift_up] and [sift_down] take
+       such a misplaced value, and move it "up" (respectively: "down")
+       until the heap ordering is restored.
+
+       Functions [sift_up] and [sift_down] do not perform swaps, but
+       rather expect the value to be assigned in the heap as an
+       additional parameter [x], resulting in twice less assignments. *)
+
+    (* store [x] at index [i], moving it up if necessary *)
+    let rec sift_up h i x =
+      if i = 0 then Dynarray.set h 0 x else
+      let p = parent_node i in
+      let y = Dynarray.get h p in
+      if E.compare x y < 0 then (
+        Dynarray.set h i y;
+        sift_up h p x
+      ) else
+        Dynarray.set h i x
+
+    let add h x =
+      let i = Dynarray.length h in
+      Dynarray.add_last h x;
+      if i > 0 then sift_up h i x
+
+    let add_iter h iter x =
+      iter (add h) x
+
+    let min_elt h =
+      if Dynarray.is_empty h then None else Some (Dynarray.get h 0)
+
+    let get_min_elt h =
+      if Dynarray.is_empty h then invalid_arg "empty priority queue";
+      Dynarray.get h 0
+
+    let lt h i j =
+      E.compare (Dynarray.get h i) (Dynarray.get h j) < 0
+
+    (* store [x] at index [i], moving it down if necessary *)
+    let rec sift_down h ~len i x =
+      let left = left_child i in
+      if left >= len then Dynarray.set h i x (* no child, stop *) else
+      let smallest =
+        let right = right_child i in
+        if right >= len then left (* no right child *) else
+        if lt h left right then left else right
+      in
+      let y = Dynarray.get h smallest in
+      if E.compare y x < 0 then (
+        Dynarray.set h i y;
+        sift_down h ~len smallest x
+      ) else
+        Dynarray.set h i x
+
+    let pop_min h =
+      let n = Dynarray.length h in
+      if n = 0 then None else
+      let x = Dynarray.pop_last h in
+      if n = 1 then Some x else (
+        let r = Dynarray.get h 0 in
+        sift_down h ~len:(n - 1) 0 x;
+        Some r
+      )
+
+    let remove_min h =
+      let n = Dynarray.length h in
+      if n > 0 then (
+        let x = Dynarray.pop_last h in
+        if n > 1 then sift_down h ~len:(n - 1) 0 x
+      )
+
+    let copy =
+      Dynarray.copy
+
+    (* array to heap in linear time (Floyd, 1964)
+
+       many elements travel a short distance, few travel longer distances
+       and we can show that it totals to O(N) *)
+    let heapify h =
+      let n = Dynarray.length h in
+      for i = n/2 - 1 downto 0 do
+        sift_down h ~len:n i (Dynarray.get h i)
+      done;
+      h
+
+    let of_array a =
+      Dynarray.of_array a |> heapify
```

Comment:
Small remark: as far as I can see, the current implement could support an "efficient" inplace map as
```ocaml
let map f x = heapify (Dynarray.map_inplace f x)
```
(if `Dynarray.map_inplace` existsted) which cannot be implemented from the interface. I don't know however if that function would be worth exposing?

Reply:
Thanks for the feedback. Indeed, we could have such a `map_inplace` (if we have `Dynarray.map_inplace`) but note that it would have the same asymptotic complexity as building a fresh array with the mapped values and then use `of_array`. It would only save us the intermediate array, and be more efficient by a small constant factor.